### ISSUE TITLE
used-before-assignment false positive when walrus / assignment expression assigned and used in same expression

### ISSUE THREAD

**Issue by OJFord** [PyCQA/pylint#3953]
Similar to #3249

### Steps to reproduce
1. `assert (n := len(stuff)) == 123, f"Expected length 123, but got {n}"`
2. `pylint`

### Current behavior
> E0601: Using variable 'n' before assignment (used-before-assignment)

### Expected behavior
Pass.

### pylint --version output
v2.6.0

**Comment by OJFord** [comment]
Another example (not using `assert`):
```python
dict({"options": (d := json.loads(data))}, data=d["data"])
```

**Comment by OJFord** [comment]
Also in ternaries, e.g.:
```python
foo = m.group("foo") if (m := FOO_PATT.match(bar)) else "bar"
```
